src/backend/distributed/relay/relay_event_utility.c

```diff
@@ -534,6 +540,33 @@ RelayEventExtendNames(Node *parseTree, char *schemaName, uint64 shardId)
 			break;
 		}
 
+		case T_SelectStmt:
+		{
+			/*
+			 * Since postgres doesn't allow custom options on a table access method table
+			 * we use a UDF to apply the options to the table. This function effectively
+			 * behaves like a ddl command.
+			 *
+			 * Since the coordinator outsourced the application of the shard id in the ddl
+			 * commands to the worker we use the same infrastructure to route this UDF
+			 * through and apply the shard name.
+			 *
+			 * Execution of the UDF will be deferred to a later point.
+			 */
+			if (IsCstoreAlterTableOptionsFunctionCall(castNode(SelectStmt, parseTree)))
+			{
+				SelectStmt *stmt = castNode(SelectStmt, parseTree);
+				RewriteCStoreAlterTableOptionsFunctionCallForShard(stmt, schemaName,
+																   shardId);
```

Comment:
In my opinion, the code generating table DDL commands is not so complex that adding this amount of undesirable code for a specific case weighs up against the cost of a (pragmatic) refactor that adapts the data model to address this new requirement. Some clever hack might also be tenable, but not sure whether they exist.

I don't think we should reinvent how DDL commands are generated, but recognize that a) DDL commands on tables are special, b) worker_apply_shard_ddl_command is a pragmatic answer to table->shard mapping but does not work in all cases.

What if the table DDL commands are all wrapped in a TableDDLCommand struct that can contain either a string or a function pointer and maybe a boolean on whether to wrap? It could also be worth introducing a TableDDLCommandList struct that contains the List to quickly catch all the callers / churn through compiler errors. 

The refactor would be replacing things like `lappend(tableDDLEventList, tableSchemaDef);` with `lappend(tableDDLEventList, NewTableDDLCommand(tableSchemaDef));` or `AppendTableDDLCommand(tableDDLEventList, tableSchemaDef)`, and adjusting the callers of `Get*TableCreationCommands` as appropriate.

We can be pragmatic here and introduce some conversion functions. It's more important for the data model to be clean than it is for code to be clean. Code eventually follows the data model.

We can leave foreign key commands and NodeDDLCommandList out of scope for now.

Reply:
I tried to make the changes in citus to facilitate this; https://github.com/citusdata/citus/compare/refactor/table-ddl-list

I feel this is indeed a better approach. Will make a change here to build on top of the above changes. Will probably add a function pointer and a `void *context` to support the columnar options.

Reply:
This is the complete integration with the alternative approach: https://github.com/citusdata/citus/compare/fix/cstore-options-dist-tables-2